session.go

```diff
@@ -633,6 +644,20 @@ func (s *session) Close() error {
 		atomic.AddInt64(&closingSessions, -1)
 		atomic.AddInt64(&openSessions, -1)
 		atomic.AddInt64(&closedSessions, 1)
+		go func() {
+			// wait until we're finished sending and receiving and then close any remaining streams
+			<-s.finishedSendingCh
+			<-s.finishedReceivingCh
+
+			s.mx.RLock()
+			for _, c := range s.streams {
+				// Note - we never send an RST because the underlying connection is
+				// considered no good at this point and we won't bother sending anything.
+				go c.close(false, nil, nil)
+			}
+			s.streams = nil
+			s.mx.RUnlock()
+		}()
```

Comment:
I got unexpected test errors when trying to remove the duplicated code in `onSessionError`. In theory code here should supersede it by making sure the send and receive loop exits before closing sessions.

Reply:
Okay have a PR https://github.com/getlantern/lampshade/pull/48 against this. Kindly let me know how you feel about that one.